packages/store/test/test-heap-classes.js

```diff
@@ -0,0 +1,107 @@
+// @ts-check
+
+import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
+import {
+  defineHeapFarClass,
+  defineHeapFarClassKit,
+  makeHeapFarInstance,
+} from '../src/patterns/interface-tools.js';
+import { M } from '../src/patterns/patternMatchers.js';
+
+const UpCounterI = M.interface('UpCounter', {
+  incr: M.call()
+    // TODO M.number() should not be needed to get a better error message
+    .optional(M.and(M.number(), M.gte(0)))
+    .returns(M.number()),
+});
+
+const DownCounterI = M.interface('DownCounter', {
+  decr: M.call()
+    // TODO M.number() should not be needed to get a better error message
+    .optional(M.and(M.number(), M.gte(0)))
+    .returns(M.number()),
+});
+
+test('test defineHeapFarClass', t => {
+  const makeUpCounter = defineHeapFarClass(
+    'UpCounter',
+    UpCounterI,
+    (x = 0) => ({ x }),
+    {
+      incr(y = 1) {
+        // @ts-expect-error TS doesn't know that `this` is a `Context`
+        const { state } = this;
+        state.x += y;
+        return state.x;
+      },
+    },
+  );
+  const upCounter = makeUpCounter(3);
+  t.is(upCounter.incr(5), 8);
+  t.is(upCounter.incr(1), 9);
+  t.throws(() => upCounter.incr(-3), {
+    message: 'In "incr" method of (UpCounter) arg 0: -3 - Must be >= 0',
+  });
+  t.throws(() => upCounter.incr('foo'), {
+    message:
+      'In "incr" method of (UpCounter) arg 0: string "foo" - Must be a number',
+  });
+});
+
+test('test defineHeapFarClassKit', t => {
+  const makeCounterKit = defineHeapFarClassKit(
+    'Counter',
+    harden({ up: UpCounterI, down: DownCounterI }),
+    (x = 0) => ({ x }),
+    {
+      up: {
+        incr(y = 1) {
+          const { state } = this;
+          state.x += y;
+          return state.x;
+        },
+      },
+      down: {
+        decr(y = 1) {
+          const { state } = this;
+          state.x -= y;
+          return state.x;
+        },
+      },
+    },
+  );
+  const { up: upCounter, down: downCounter } = makeCounterKit(3);
+  t.is(upCounter.incr(5), 8);
+  t.is(downCounter.decr(), 7);
+  t.is(upCounter.incr(3), 10);
+  t.throws(() => upCounter.incr(-3), {
+    message: 'In "incr" method of (Counter up) arg 0: -3 - Must be >= 0',
+  });
+  // @ts-expect-error the type violation is what we're testing
+  t.throws(() => downCounter.decr('foo'), {
+    message:
+      'In "decr" method of (Counter down) arg 0: string "foo" - Must be a number',
+  });
+  t.throws(() => upCounter.decr(3), {
```

Comment:
Since you didn't suppress a type warning, I take it signature types don't propagate through?

Reply:
I'm sorry, I'm just baffled by the types. My approach is to keep them as precise as I know how, but when I don't know how to just proceed anyway. I do not have a model of when to expect it to infer what precision of types.

@turadg @mhofman @michaelfig ?

Reply:
I've been surprised by this case too. The type of `upCounter` does have the signature:
<img width="473" alt="Screen Shot 2022-09-04 at 7 46 51 AM" src="https://user-images.githubusercontent.com/21505/188319451-9bc25206-750c-4130-8479-d4bf4797bbfa.png">

But the type inference is allowing access of unknown properties:
<img width="185" alt="Screen Shot 2022-09-04 at 7 54 53 AM" src="https://user-images.githubusercontent.com/21505/188319827-e69ddce5-1af4-4e7c-85fc-416831efbadf.png">

That's because we allow implicit any. With [noImplicitAny](https://www.typescriptlang.org/tsconfig#noImplicitAny) enabled:
<img width="622" alt="Screen Shot 2022-09-04 at 8 01 40 AM" src="https://user-images.githubusercontent.com/21505/188320077-b5727126-d783-4884-a4e7-2393d62970a2.png">

That's not feasible because it would raise tons of type warnings. Declaring a type (skipping inference) solves the `decr` case, but the declaration doesn't know what `this` will be:
<img width="426" alt="Screen Shot 2022-09-04 at 8 09 10 AM" src="https://user-images.githubusercontent.com/21505/188320453-77639fcc-7919-4998-9b10-758f904654ba.png">

Example when it does:
<img width="465" alt="Screen Shot 2022-09-04 at 8 11 00 AM" src="https://user-images.githubusercontent.com/21505/188320520-32c550d1-380b-4fee-9571-01cc63c43753.png">